### PR TITLE
Docs: Update Quaternion.html

### PR DESCRIPTION
**Pull request by OmarShehata**
### Summary

The [documentation page for quaternions](https://threejs.org/docs/#api/en/math/Quaternion) states that one of the reasons quaternions are used is to avoid gimbal lock. This is a common misconception. Gimbal lock will happen when describing rotations with Euler angles, regardless of whether these angles are composed together as 3 matrices, or 3 quaternions, or something else.

Gimbal lock is resolved when using an axis-angle system and composing rotations by multiplying into a stored quaternion (or matrix).

This PR fixes the doc by instead highlighting one of the reasons quaternions are typically used (less numbers required for storage), which is good because composing rotations requires less operations and is thus computationally faster.

### Why this matters

Quaternions are already somewhat intimidating due to requiring more mathematical background than you might usually need to understand how and why they work. I believe keeping this misconception in the doc further adds to any confusion the user may have, and at best obscures the real reasons why engines use quaternions (faster to compose, easier to interpolate).

This misconception may cause users to re-implement their systems to use quaternions when encountering gimbal lock, only to discover that that was not necessary to fix it nor sufficient, and that an axis-angle system with matrices may be perfectly fine and would have saved them a lot of time (this is exactly what happened to me a few years ago and hence why I'm here).

### Details

I want to start off by saying I'm not here to make any claims about the merits of quaternions and whether to encourage or dissuade people from using them. I simply want to be clear about (1) what exactly is it that quaternions do that you can't do with other systems of describing rotations and (2) what exactly causes gimbal lock and how to fix it.

Most of what I wanted to say here has been eloquently stated 20 years ago in the [gamdev.net discussion](https://www.gamedev.net/articles/programming/math-and-physics/do-we-really-need-quaternions-r1199) linked to in the doc here:

https://threejs.org/docs/index.html#api/en/math/Matrix4.makeRotationAxis

_Note that this doc states that constructing axis angles from matrices is a mathematically sound alternative to rotating with quaternions, which supports my point that gimbal lock is not uniquely solved by using quaternions._

The discussion above explains how you can use rotation matrices to get identical results to a system using quaternions (including creating a system that does not suffer from gimbal lock). The conclusion from that post is that we don't need quaternions. I believe there are valid reasons to use quaternions over matrices:

* Eric Haines explains [in this video](https://www.youtube.com/watch?v=Ocoibc7MoKg) that interpolating rotations correctly with axis-angle is much easier than with Euler angles, that this is simple to do with quaternions and that this simplicity is why many engines use quaternions. 
* Eric Lengyel's _Fundamentals of Game Engine Design_ volume 1 page 89 explains that even though rotating a vector by a quaternion actually takes more operations than rotating by a 3x3 matrix, composing quaternions together only takes 16 multiplies and 12 adds, compared to 27 multiplies and 18 adds required to compute the product of 3x3 matrices, so that gives you a computational speed up).

These are two good reasons to choose quaternions for your rotation system. Resolving gimbal lock is not one of them. In fact, a rotation system that uses quaternions is no more and no less susceptible to the issue of gimbal lock than one that uses matrices. It doesn't matter what method you're using to describe, store and compose your rotations. It matters whether you're using a system of 3 fixed angles (Euler) or a system that lets you accumulate rotations of arbitrary axis-angles.

If you're not convinced by the linked gamedev.net discussion, I put together this demo as a way to clearly show how a system with quaternions can gimbal lock:

https://gimbal-lock-quaternions.glitch.me/

Controls:

* Rotate the cubes with W,A,S,D and Q,E.
* Reset with R

_(doesn't really work on mobile, sorry)._ 

The easiest way to get the cubes to gimbal lock here is to hold D until you see the **-X** side head on. The top two scenes are now in a locked state. Notice that **W/S** and **Q/E** are now both rotating the cube in the same way. You can no longer rotate around the Z axis with Q/E (pointing perpendicular from the screen) as shown by the bottom two scenes which do not experience gimbal lock.

You can inspect the code for this here:

https://glitch.com/edit/#!/gimbal-lock-quaternions?path=index.html:1:0

Click on **index.html** on the left. Each of the 4 rotation systems is documented here. Towards the end is the rotation system that uses quaternions but still has gimbal lock. This essentially creates 3 quaternions, each has a set angle that is controlled by the keyboard, and those 3 quaternions are composed in a specific order before being set to the cube's rotation.

You might say this is "using quaternions wrong", but why do we not say that about matrices when used in that same way? The answer is that the choice to use quaternions vs matrices tells you nothing about whether a system will gimbal lock. It depends on how they're composed. It is therefore incorrect for the doc to state that quaternions are used to avoid gimbal lock.

### In Conclusion

I am happy to hear from the experts in this community about whether you agree that this is indeed a misconception that would be helpful to remove. I appreciate your time on this very lengthy PR for what is a single line doc change.

I hope the linked demo and code help untangle the ideas of quaternions vs matrices vs euler angles vs axis angles, if anyone has struggled with this as I have.

### CODE REVIEW

**Comment by looeee** [comment]
> Most of what I wanted to say here has been eloquently stated 20 years ago in the gamdev.net discussion linked to in the doc here:

Did you arrive at that link by way of the [hackernews post](https://news.ycombinator.com/item?id=18365433) from last week? The discussion on that gamedev post was quite the rollercoaster 😅 

Regarding the docs: I'm not sure that we should explain anything about the reasons for choosing quaternions/euler angles/rotation matrix on the quaternion doc page. We should certainly remove the note about gimbal lock, but I'm not convinced that your sentence is that much of an improvement. The reasons for choosing one or the other representation is too nuanced to explain in a single sentence. 

If we do attempt to explain things, we should do so fully, and that would work best as a separate manual page - maybe **Representing 3D rotations** - to go alongside the existing [Matrix transformations](https://threejs.org/docs/#manual/en/introduction/Matrix-transformations) page.

In the meantime, perhaps we could provide users with a link to an introductory article?

**Comment by Mugen87** [comment]
Somebody should ping Unity so they correct their docs: https://docs.unity3d.com/ScriptReference/Quaternion.html

> Quaternions are used to represent rotations. They are compact, don't suffer from gimbal lock and can easily be interpolated. Unity internally uses Quaternions to represent all rotations.

**Comment by looeee** [comment]
> The conclusion from that post is that we don't need quaternions.

Many people would not agree with that conclusion - there's even a big warning at the top of the post saying that the the gamedev.net staff believe it's incorrect and only keep it there for historical purposes. 

I actually thought you had linked to this rebuttal thread, which is where the really interesting discussion on this topic happens:

https://www.gamedev.net/forums/topic.asp?topic_id=25314

> gimbal lock is not uniquely solved by using quaternions.

Not _uniquely_ solved - rotation matrices also solve this problem, as do [rotors](https://en.wikipedia.org/wiki/Rotor_(mathematics)) and probably many other ways of representing rotations. But there was no claim that quaternions are the only way to solve this problem. 

I think the original claim on the quaternions page is intended to compare them to Euler angles - in other words, it's intended to convey something like this to people who don't know a lot about 3D rotations and are wondering what quaternions are all about:

_Even though quaternions are more difficult to use than Euler angles, they don't suffer from gimbal lock which makes the extra difficulty of using them worthwhile._ 

But I think the gimbal lock thing is probably the least interesting reason for using quaternions over Euler angles. Much more important are composition and interpolation.

Maybe we could put something like this as the intro to the quaternions page:

Quaternions have the following advantages over Euler angles:

1. They don't suffer from Gimbal lock
2. It's possible to compose quaternions
3. It's possible to interpolate between quaternions which allows for animating rotation

**Comment by Mugen87** [comment]
I really don't think that forum discussions are a good and reliable source for facts. It's not surprising that they are no valid references for any form of scientific work.

 An established and well-known literature about math for graphics and game development is: _3D Math Primer for Graphics and Game Development_ by Fletcher Dunn and Ian Parberry.

In chapter 8.5, the say about quaternions:

> There is a mathematical reason why using only three numbers to represent a 3-space orientation is guaranteed to cause the problems we discussed with Euler angles, such as Gimbal lock. It has something to do with some fairly advanced math terms such as “manifolds.” A quaternion avoids these problems by using four numbers to express an orientation (hence the name quaternion).

**Comment by OmarShehata** [comment]
> Did you arrive at that link by way of the hackernews post from last week? (EDIT: I mean this hackernews post) The discussion on that gamedev post was quite the rollercoaster 😅

I actually only discovered that gamedev.net discussion when putting together my rotation demo up there, when showing that you can use axis-angle with matrices, it's linked to in the ThreeJS doc here: https://threejs.org/docs/index.html#api/en/math/Matrix4.makeRotationAxis

> We should remove the note about gimbal lock, but I'm not convinced that your sentence is that much of an improvement. The reasons for choosing one or the other representation is too nuanced to explain in a single sentence.

I do agree with this that this may still be confusing if it doesn't link to somewhere you can more fully understand the reasons. Since the doc already links to the Wikipedia page on rotations with quaternions, I've removed the justification from the sentence.

> I think the original claim on the quaternions page is intended to compare them to Euler angles

I did consider this, and I wondered if just adding "as opposed to used Euler angles" or something like that would fix it, but I still have mixed feelings about the unconditional phrase "quaternions don't suffer from gimbal lock". I guess one thing I don't quite fully understand yet is: is there anything inherently wrong with using quaternions to represent euler angles? It's definitely unconventional, and maybe it defeats the whole purpose they were created for (I don't know)? **Would we find the sentence "Rotation matrices don't suffer from gimbal lock" accurate?** Or is it just because there isn't a convention about using rotation matrices with axis-angle accumulating rotations that this sentence seems inaccurate/requires qualification?

 In any case, I've said my piece. I appreciate you all being willing to change the doc to clear any confusion. Feel free to tweak any wording in this PR or do what you will with it.

**Review by Mugen87 — APPROVED** [review]

File: docs/api/en/math/Quaternion.html [hunk 1]
```diff
@@ -12,10 +12,7 @@ <h1>[name]</h1>
 
 		<p class="desc">
 			Implementation of a [link:http://en.wikipedia.org/wiki/Quaternion quaternion].
-			This is used for [link:https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation rotating things]
-			without encountering the dreaded
-			[link:http://en.wikipedia.org/wiki/Gimbal_lock gimbal lock] issue, amongst other
-			advantages.
+			This is used for [link:https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation rotating things].
```
Comment (WestLangley) [inline docs/api/en/math/Quaternion.html]: Instead, just say

>Quaternions are used in three.js to represent rotations.

**Comment by WestLangley** [comment]
For those interested, have a look at this post -- in particular, the answer: https://mathoverflow.net/questions/95902/the-gimbal-lock-shows-up-in-my-quaternions

I agree with what @Mugen87 said that Unity said:   :-)
>Quaternions ... are compact, don't suffer from gimbal lock and can easily be interpolated.

File: docs/api/en/math/Quaternion.html (hunk 1 above)
Comment (mrdoob) [reply docs/api/en/math/Quaternion.html]: Like this?
```
Quaternions are used to [link:https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation represent rotations].
```

**Comment by peteroupc** [comment]
See also this [Stack Overflow question](https://stackoverflow.com/questions/28776788), which shows that quaternions can still cause gimbal locks when they're used merely as a substitute for pitch/yaw/roll angles.

File: docs/api/en/math/Quaternion.html (hunk 1 above)
Comment (WestLangley) [reply docs/api/en/math/Quaternion.html]: Like this:
```html
Implementation of a [link:http://en.wikipedia.org/wiki/Quaternion quaternion].
Quaternions are used in three.js to represent [link:https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation rotations].
```

Comment (mrdoob) [reply docs/api/en/math/Quaternion.html]: Sounds good. Will change post-merge.

**Comment by mrdoob** [comment]
Thanks!
